## Assignment-B Multivariant Linear Regression
@author: Kai-Ping Wang

## Importing the libraries
Import all required libraries, and also set some libraries options.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option('display.max_columns', None)

## Importing the dataset

In [2]:
file_url = 'https://raw.githubusercontent.com/aso-uts/applied_ds/master/assignment1/cancer_reg.csv'
df = pd.read_csv(file_url, encoding='ISO-8859-1')

## Checking the dataset
Check the data structure and information to be sure if data processing is required

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3047 entries, 0 to 3046
Data columns (total 34 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   avgAnnCount              3047 non-null   float64
 1   avgDeathsPerYear         3047 non-null   int64  
 2   TARGET_deathRate         3047 non-null   float64
 3   incidenceRate            3047 non-null   float64
 4   medIncome                3047 non-null   int64  
 5   popEst2015               3047 non-null   int64  
 6   povertyPercent           3047 non-null   float64
 7   studyPerCap              3047 non-null   float64
 8   binnedInc                3047 non-null   object 
 9   MedianAge                3047 non-null   float64
 10  MedianAgeMale            3047 non-null   float64
 11  MedianAgeFemale          3047 non-null   float64
 12  Geography                3047 non-null   object 
 13  AvgHouseholdSize         3047 non-null   float64
 14  PercentMarried          

In [4]:
df.describe(include='all')

avgAnnCount  avgDeathsPerYear  TARGET_deathRate  incidenceRate  \
count    3047.000000       3047.000000       3047.000000    3047.000000   
unique           NaN               NaN               NaN            NaN   
top              NaN               NaN               NaN            NaN   
freq             NaN               NaN               NaN            NaN   
mean      606.338544        185.965868        178.664063     448.268586   
std      1416.356223        504.134286         27.751511      54.560733   
min         6.000000          3.000000         59.700000     201.300000   
25%        76.000000         28.000000        161.200000     420.300000   
50%       171.000000         61.000000        178.100000     453.549422   
75%       518.000000        149.000000        195.200000     480.850000   
max     38150.000000      14010.000000        362.800000    1206.900000   

            medIncome    popEst2015  povertyPercent  studyPerCap  \
count     3047.000000  3.047000e+03     3047.000000  3047.000000   
unique            NaN           NaN             NaN          NaN   
top               NaN           NaN             NaN          NaN   
freq              NaN           NaN             NaN          NaN   
mean     47063.281917  1.026374e+05       16.878175   155.399415   
std      12040.090836  3.290592e+05        6.409087   529.628366   
min      22640.000000  8.270000e+02        3.200000     0.000000   
25%      38882.500000  1.168400e+04       12.150000     0.000000   
50%      45207.000000  2.664300e+04       15.900000     0.000000   
75%      52492.000000  6.867100e+04       20.400000    83.650776   
max     125635.000000  1.017029e+07       47.400000  9762.308998   

                 binnedInc    MedianAge  MedianAgeMale  MedianAgeFemale  \
count                 3047  3047.000000    3047.000000      3047.000000   
unique                  10          NaN            NaN              NaN   
top     (54545.6, 61494.5]          NaN            NaN              NaN   
freq                   306          NaN            NaN              NaN   
mean                   NaN    45.272333      39.570725        42.145323   
std                    NaN    45.304480       5.226017         5.292849   
min                    NaN    22.300000      22.400000        22.300000   
25%                    NaN    37.700000      36.350000        39.100000   
50%                    NaN    41.000000      39.600000        42.400000   
75%                    NaN    44.000000      42.500000        45.300000   
max                    NaN   624.000000      64.700000        65.700000   

                            Geography  AvgHouseholdSize  PercentMarried  \
count                            3047       3047.000000     3047.000000   
unique                           3047               NaN             NaN   
top     Charles City County, Virginia               NaN             NaN   
freq                                1               NaN             NaN   
mean                              NaN          2.479662       51.773679   
std                               NaN          0.429174        6.896928   
min                               NaN          0.022100       23.100000   
25%                               NaN          2.370000       47.750000   
50%                               NaN          2.500000       52.400000   
75%                               NaN          2.630000       56.400000   
max                               NaN          3.970000       72.500000   

        PctNoHS18_24   PctHS18_24  PctSomeCol18_24  PctBachDeg18_24  \
count    3047.000000  3047.000000       762.000000      3047.000000   
unique           NaN          NaN              NaN              NaN   
top              NaN          NaN              NaN              NaN   
freq             NaN          NaN              NaN              NaN   
mean       18.224450    35.002068        40.977034         6.158287   
std         8.093064     9.069722        11.115805         4.52

## Data Processing
As we are using only numeric features, we will drop out Geography and binnedInc columns.

In [5]:
df_cleaned = df.copy()
df_cleaned = df_cleaned.drop(columns=['Geography','binnedInc'])

## Run Linear Regression
Run the univariant linear regression on state against TARGET_deathRate. 

Then print out the score and MSE, and also draw the diagram of ground truth and prediction model.

### Create getScore(X,y) method for reuse
Set up a method to split the preprocessed data into training (80%) and testing (20%) set, and then run linear regression on the training set while printing out the R Square score and MSE using the testing set.

In [6]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error as mse
from sklearn.model_selection import train_test_split

def getScore(X,y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
    reg = LinearRegression()
    reg.fit(X_train,y_train)
    print("Score is %.4f" % reg.score(X_test,y_test))
    print("Training set MSE is %.4f" % mse(y_train, reg.predict(X_train)))
    print("Testing set MSE is %.4f" % mse(y_test, reg.predict(X_test)))
    return reg.score(X_test,y_test)


### Drop all missing data and try to see the result
See the baseline without any further pre-processing

In [7]:
df_proc = df_cleaned.dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']

getScore(X,y)

Score is 0.3013
Training set MSE is 363.2677
Testing set MSE is 529.6852


0.30125046110409215

### Clean up unreasonable data
By quickly review the data, we noticed that 
* MedianAge has number larger than 100, which is probably a typo and should be divided by 10. 
* AvgHouseholdSize has value smaller than 1, which is probably a typo and should be multiplied by 100.

In [8]:
df_cleaned.AvgHouseholdSize = df_cleaned.AvgHouseholdSize.apply(lambda x: x if x > 1 else x*100)
df_cleaned.MedianAge = df_cleaned.MedianAge.apply(lambda x: x if x < 100 else x/10)

### Rerun the baseline
The expectation here is the result should be better as included data is more accurate and meaningful.

In [9]:
df_proc = df_cleaned.dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']

getScore(X,y)

Score is 0.3066
Training set MSE is 364.7068
Testing set MSE is 525.6571


0.30656420507535764

### Too many missing data for feature - PctSomeCol18_24

In [10]:
# PctSomeCol18_24 only has 25% meaningful data. Try Removing this column completely.
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns='PctSomeCol18_24').dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4490
Training set MSE is 367.9426
Testing set MSE is 416.9053


0.44900944584382485

Try manipulate PctSomeCol18_24 to use average for missing data

Define replaceMissingWithMean(df, feature) for reuse

In [11]:
def replaceMissingWithMean(df, feature):
    mean = df.dropna(how='any', subset=[feature])[feature].mean()
    df[feature].fillna(value=mean, inplace=True)
    df.dropna(how='any', inplace=True)
    return df

In [12]:
# Try replacing missing value with mean
df_proc = df_cleaned.copy()
df_proc = replaceMissingWithMean(df_proc, 'PctSomeCol18_24')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4487
Training set MSE is 367.9406
Testing set MSE is 417.1099


0.44873907546643754

Define replaceMissingWithMedian(df, feature) for reuse

In [13]:
def replaceMissingWithMedian(df, feature):
    median = df.dropna(how='any', subset=[feature])[feature].median()
    df[feature].fillna(value=median, inplace=True)
    df.dropna(how='any', inplace=True)
    return df

In [14]:
# Try replacing missing value with median
df_proc = df_cleaned.copy()
df_proc = replaceMissingWithMedian(df_proc, 'PctSomeCol18_24')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4487
Training set MSE is 367.9394
Testing set MSE is 417.1596


0.44867337554465336

The performance is about the same between deleting this feature and replacing with mean or median value. Remove this feature

### Handle next feature with missing value - PctPrivateCoverageAlone

In [15]:
# Try removing the feature completely and see the performance
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone']).dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4869
Training set MSE is 365.7441
Testing set MSE is 384.5184


0.48686780406878805

In [16]:
# Try replacing missing value with mean
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24'])
df_proc = replaceMissingWithMean(df_proc, 'PctPrivateCoverageAlone')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4873
Training set MSE is 365.7100
Testing set MSE is 384.2174


0.4872695535129641

In [17]:
# Try replacing missing value with median
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24'])
df_proc = replaceMissingWithMedian(df_proc, 'PctPrivateCoverageAlone')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.4873
Training set MSE is 365.7098
Testing set MSE is 384.2158


0.48727161448240575

The performance is about the same between deleting this feature and replacing with mean or median value.

### Handle next feature with missing value - PctEmployed16_Over

In [18]:
# Try removing the feature completely
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over']).dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5640
Training set MSE is 376.0007
Testing set MSE is 343.9928


0.5639999062278332

In [19]:
# Replacing missing value with mean
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone'])
df_proc = replaceMissingWithMean(df_proc, 'PctEmployed16_Over')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5636
Training set MSE is 372.6659
Testing set MSE is 344.2809


0.5636348386287059

In [20]:
# Replacing missing value with median
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone'])
df_proc = replaceMissingWithMedian(df_proc, 'PctEmployed16_Over')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5636
Training set MSE is 372.6756
Testing set MSE is 344.2759


0.5636410786374628

The performance is about the same between deleting this feature and replacing with mean value.

### Lots of 0 value in feature - studyPerCap
Although studyPerCap feature does not have missing value, there are lots of 0 in this feature. Treat the value 0 as missing value. Only about 30% of data has meaningful value for this feature.

In [21]:
df_proc[df_proc.studyPerCap != 0].shape

(1116, 30)

In [22]:
# Remove this feature completely
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap']).dropna(how='any')
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5640
Training set MSE is 376.0020
Testing set MSE is 343.9717


0.5640266487236938

In [23]:
# Try replacing 0 with mean
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over']).dropna(how='any')

mean = df_proc[df_proc.studyPerCap != 0].studyPerCap.mean()
df_proc.studyPerCap.replace(to_replace=0, value=mean, inplace=True)
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5642
Training set MSE is 375.9639
Testing set MSE is 343.8452


0.5641870032538023

In [24]:
# Try dropping missing value rows
df_proc = df_cleaned.copy()
df_proc = df_proc.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over']).dropna(how='any')
df_proc = df_proc[df_proc.studyPerCap != 0]
X = df_proc.drop(columns='TARGET_deathRate').values
y = df_proc['TARGET_deathRate']
getScore(X,y)

Score is 0.5226
Training set MSE is 190.7173
Testing set MSE is 282.8436


0.5226133010304227

The performance has greatly improved, but it can be due to the much lesser amount of data. It is something worth exploring in the future.

### Lots of duplicate value in feature - avgAnnCount
Although avgAnnCount feature does not have missing value, there are lots of 1962.667684 in this feature. Treat the value 1962.667684 as missing value. Around 30% of data has meaningful avgAnnCount.

In [25]:
df_proc[df_proc.avgAnnCount != 1962.667684].shape

(1058, 29)

In [26]:
# Removing those missing value rows
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap']).dropna(how='any')
df_dropna = df_dropna[df_dropna.avgAnnCount != 1962.667684]
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)

Score is 0.5432
Training set MSE is 356.9329
Testing set MSE is 313.1803


0.5432403559785393

In [27]:
# Delete this feature completely
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap','avgAnnCount']).dropna(how='any')
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)

Score is 0.5609
Training set MSE is 378.5052
Testing set MSE is 346.4604


0.5608723551353566

The performance is about the same

### Lot of value 453.5494221 for feature - incidenceRate


In [28]:
# Removing those missing rows
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap','avgAnnCount']).dropna(how='any')
df_dropna = df_dropna[df_dropna.incidenceRate != 453.5494221]
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)

Score is 0.5109
Training set MSE is 370.8431
Testing set MSE is 335.3327


0.5109320689101766

In [29]:
# Try removing this feature completely
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap','avgAnnCount','incidenceRate']).dropna(how='any')
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)

Score is 0.4254
Training set MSE is 456.7620
Testing set MSE is 453.3053


0.4254497990969408

The performance has dropped significantly, so this feature is quite important in the prediction model.

In [30]:
# Try replacing the dummy value with the mean
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','studyPerCap','avgAnnCount']).dropna(how='any')
mean = df_dropna[df_dropna.incidenceRate != 453.5494221].incidenceRate.mean()
df_dropna['incidenceRate'].replace(to_replace=453.5494221, value=mean, inplace=True)
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)


Score is 0.5615
Training set MSE is 378.1749
Testing set MSE is 346.0027


0.5614524339573921

The final value has been consistent across previous cells

Score is 0.5615

Training set MSE is 378.1749

Testing set MSE is 346.0027

In [47]:
from sklearn.ensemble import RandomForestRegressor

def getScoreRT(X,y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
    reg = RandomForestRegressor(n_estimators = 100, max_depth=5, min_samples_leaf=2, min_samples_split=2, random_state = 0)
    reg.fit(X_train,y_train)
    print("Score is %.4f" % reg.score(X_test,y_test))
    print("Training set MSE is %.4f" % mse(y_train, reg.predict(X_train)))
    print("Testing set MSE is %.4f" % mse(y_test, reg.predict(X_test)))
    return reg.score(X_test,y_test)
getScoreRT(X,y)


Score is 0.5238
Training set MSE is 300.6347
Testing set MSE is 375.7230


0.5237828825983781

In [31]:
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators = 10, random_state = 0)
regressor.fit(X, y)

RandomForestRegressor(n_estimators=10, random_state=0)

In [36]:
getScoreRT(X,y)

Score is 0.5118
Training set MSE is 72.2623
Testing set MSE is 385.1399


0.5118473069868028

In [60]:
from sklearn.preprocessing import StandardScaler
y_array = y.values
y_2d = y_array.reshape(len(y_array),1)
sc_X = StandardScaler()
sc_y = StandardScaler()
Xsc = sc_X.fit_transform(X)
ysc = sc_y.fit_transform(y_2d)


In [61]:
from sklearn.svm import SVR

def getScoreSVM(X,y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
    reg = SVR(kernel = 'rbf')
    reg.fit(X, y)
    reg.fit(X, y)
    print("Score is %.4f" % reg.score(X_test,y_test))
    print("Training set MSE is %.4f" % mse(y_train, reg.predict(X_train)))
    print("Testing set MSE is %.4f" % mse(y_test, reg.predict(X_test)))
    return reg.score(X_test,y_test)
getScoreSVM(Xsc,ysc)


Score is 0.6489
Training set MSE is 0.3251
Testing set MSE is 0.3598


0.6489281691761573

In [37]:
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over'])
features = df_dropna.columns

for f in features:
    if f != 'TARGET_deathRate':
        df_dropna = df_cleaned.copy()
        df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over',f]).dropna(how='any')
        X = df_dropna.drop(columns='TARGET_deathRate').values
        y = df_dropna['TARGET_deathRate']
        score = getScore(X,y)
        if score >= 0.5650:
            print(f)

Score is 0.5608
Training set MSE is 378.4930
Testing set MSE is 346.5180
Score is 0.5622
Training set MSE is 379.1588
Testing set MSE is 345.3914
Score is 0.4261
Training set MSE is 455.6572
Testing set MSE is 452.8194
Score is 0.5648
Training set MSE is 376.6097
Testing set MSE is 343.3886
Score is 0.5650
Training set MSE is 377.3284
Testing set MSE is 343.1759
popEst2015
Score is 0.5630
Training set MSE is 377.7241
Testing set MSE is 344.7507
Score is 0.5640
Training set MSE is 376.0020
Testing set MSE is 343.9717
Score is 0.5642
Training set MSE is 376.0308
Testing set MSE is 343.8681
Score is 0.5639
Training set MSE is 376.2066
Testing set MSE is 344.0774
Score is 0.5640
Training set MSE is 376.0008
Testing set MSE is 343.9650
Score is 0.5591
Training set MSE is 377.0097
Testing set MSE is 347.8372
Score is 0.5640
Training set MSE is 378.0870
Testing set MSE is 343.9797
Score is 0.5624
Training set MSE is 376.2795
Testing set MSE is 345.2868
Score is 0.5575
Training set MSE is 378.

In [38]:
# Try replacing the dummy value with the mean
df_dropna = df_cleaned.copy()
df_dropna = df_dropna.drop(columns=['PctSomeCol18_24','PctPrivateCoverageAlone','PctEmployed16_Over','avgAnnCount','avgDeathsPerYear','medIncome','popEst2015','PctPublicCoverage','PctPublicCoverageAlone','PctMarriedHouseholds']).dropna(how='any')
X = df_dropna.drop(columns='TARGET_deathRate').values
y = df_dropna['TARGET_deathRate']
getScore(X,y)


Score is 0.5660
Training set MSE is 382.2834
Testing set MSE is 342.4393


0.565968989229412